sdks/go/pkg/beam/core/graph/coder/bytes_test.go

```diff
@@ -0,0 +1,61 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package coder
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestEncodeDecodeBytes(t *testing.T) {
+	longString := strings.Repeat(" this sentence is 32 characters.", 8) // 256 characters to ensure LP works.
+	tests := []struct {
+		v       []byte
+		encoded []byte
+	}{
+		{v: []byte{}, encoded: []byte{0}},
+		{v: []byte{42}, encoded: []byte{1, 42}},
+		{v: []byte{42, 23}, encoded: []byte{2, 42, 23}},
+		{v: []byte(longString), encoded: append([]byte{128, 2}, []byte(longString)...)},
```

Comment:
In order for the VarInt encoding to work, it sacrifices 1 bit out of 8 to signal "I'm not done yet". So the first one is set to 128 to signal I have more dat, and the last byte in the sequence has the leading bit 0 to say I'm done. You can read a bit about how the Protocol Buffer Varint encoding works (which is the same as what Beam uses) here: https://carlmastrangelo.com/blog/lets-make-a-varint

To be more specific. 256 in binary is 0b1 0000 0000  (9 bits). Since we only have 7 bits per byte available, due to the continuation bit, it needs to be spread over 2 bytes: so it gets encoded as 0b1000 0000 (just the continue bit) + 0b0000 0010 (2)

Reply:
Ah gotcha. Makes sense now.